### PR TITLE
feat: grant ownership follow-up

### PR DESCRIPTION
**Pull request by sfc-gh-jcieslak**
A follow-up for https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/2604.

Done in this pr:
- All of the edge cases handled and tested (except of tasks that are done in the separate pr):
  - Materialized views (already handled by Snowflake no changes needed)
  - RBAC hierarchy (test case added)
  - Delete dependent resource (role or granted object) and remove grant resource from the state (test case added)

Won't do:
- External tables (cannot handle this edge case, because we have to know the auto_refresh state of the external table; it's not retrievable by SHOW or DESC commands. It will be still possible to grant ownership of the external table, but there may be additional manual work to do afterward. Everything is documented.)

## Test Plan

* [x] acceptance tests that show how the resource is handling certain edge cases + RBAC use case

## References
[GRANT OWNERSHIP](https://docs.snowflake.com/en/sql-reference/sql/grant-ownership)

## Mentioned in
A list of issues requesting this resource: https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2549 https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2199 https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2084 https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/1942 https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/1875

### CODE REVIEW

File: docs/resources/grant_ownership.md [hunk 1]
```diff
@@ -1,15 +1,237 @@
...
+- OPERATE and MONITOR privileges are granted to the current role on the pipe(s) and `outbound_privileges` field is set to `COPY`.
+- The pipe(s) running status is paused (additional privileges and fields set are needed to pause and resume the pipe before and after ownership transfer. If it's already paused, nothing additional is needed and the pipe will remain paused after the ownership transfer).
+
+To transfer ownership of a pipe(s) **semi-automatically** you have to:
+1. Pause the pipe(s) you want to transfer ownership of (using [ALTER PIPE](https://docs.snowflake.com/en/sql-reference/sql/alter-pipe#syntax); see PIPE_EXECUTION_PAUSED).
+2. Create Terraform configuration with the `snowflake_grant_ownership` resource and perform ownership transfer with the `terraform apply`.
+3. To resume the pipe(s) after ownership transfer use [PIPE_FORCE_RESUME system function](https://docs.snowflake.com/en/sql-reference/functions/system_pipe_force_resume).
+
+## Granting ownership on tasks
+
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply docs/resources/grant_ownership.md:229]: Commented

File: pkg/resources/grant_ownership_acceptance_test.go [hunk 2]
```diff
@@ -579,6 +556,245 @@ func TestAcc_GrantOwnership_InvalidConfiguration_MultipleTargets(t *testing.T) {
...
+					PostApplyPostRefresh: []plancheck.PlanCheck{
+						plancheck.ExpectEmptyPlan(),
+					},
+				},
+			},
+		},
+	})
+}
+
+func roleBasedAccessControlUseCaseConfig(accountRoleName string, databaseName string, userName string, schemaName string, withSecondaryProvider bool) string {
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_acceptance_test.go:753]: Hmm, I agree that splitting those functions is better, but in this case, I think I would leave it as is. It's only used in one test case and splitting wouldn't be that beneficial, because the `b()` config is strongly connected to the `a()` one, so It wouldn't make much sense to use them separately.

File: pkg/resources/grant_ownership_acceptance_test.go [hunk 3]
```diff
@@ -850,3 +1066,54 @@ func checkResourceOwnershipIsGranted(opts *sdk.ShowGrantOptions, grantOn sdk.Obj
...
+	ctx := context.Background()
+	roleId := sdk.NewAccountObjectIdentifier(name)
+	assert.NoError(t, client.Roles.Create(ctx, sdk.NewCreateRoleRequest(roleId)))
+
+	return func() {
+		assert.NoError(t, client.Roles.Drop(ctx, sdk.NewDropRoleRequest(roleId)))
+	}
+}
+
+func createDatabase(t *testing.T, name string) func() {
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_ownership_acceptance_test.go:1084]: Refactored, so grant ownership specific things are in a separate function

File: pkg/resources/grant_ownership_acceptance_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/resources/grant_ownership_acceptance_test.go:759]: I disagree, this will cause more similar non-reusable functions to arise and it could be split in to a few reusable ones.

Helpers rework is incoming, though, so we can leave this as is for now. Resolve after reading.

**Review by sfc-gh-asawicki — APPROVED** [review]

**Review by sfc-gh-asawicki — APPROVED** [review]
